contracts/Voting.sol

```diff
@@ -0,0 +1,141 @@
+        string[] memory options;
+        options[0] = "Yes";
+        options[1] = "No";
+        return startPoll(_description, _allowMultipleOptions, options);
+    }
+
+    function endPoll(uint256 _pollID) external onlyOwner {
+        Poll storage poll = polls[_pollID];
+        poll.status = Status.CLOSED;
+        returnCoinsAfterPoll(_pollID);
+    }
+
+    function castVote(uint256 _pollID, uint256 _optionIndex, uint256 weight) public {
+        require(getStatus(_pollID) == Status.IN_PROGRESS, "Poll is not in progress!");
+        require(weight > 0, "Cannot vote with 0 tokens!");
+        address voter = msg.sender;
+        // requires approval from the voter beforehand to transfer coins from his/her account to this contract owner address
+        // to do that call the approve-method in the ERC20-Token before casting the vote
+        token.transferFrom(voter, owner, weight);
+        Poll storage poll = polls[_pollID];
+        // add vote
+        poll.votes[voter][_optionIndex] += weight;
+        // if address has not voted before, push to list of voters
+        if (poll.addressToTotalVotes[voter] == 0) {
+            poll.voters.push(voter);
+        }
+        // update total vote acount
+        poll.addressToTotalVotes[voter] += weight;
+    }
+
+    function getStatus(uint256 _pollID) view public returns (Status) {
+        return polls[_pollID].status;
+    }
+
+    function getResults(uint256 _pollID) view public returns(uint256[] memory){
+        Poll storage poll = polls[_pollID];
+        uint256[] memory result = new uint256[](poll.votingOptionsCount);
+
+        // iterate through the voting options
+        for(uint256 votingOptionIndex; votingOptionIndex < poll.votingOptionsCount; votingOptionIndex++) {
+            // initialize vote counter
+            uint256 amountVotes;
+            // iterate through the votes everyone has sent
+            for(uint voter = 0; voter < poll.voters.length; voter++) {
+                address voterAddress = poll.voters[voter];
+                // add votes for that voting option to the votes counter
+                amountVotes += poll.votes[voterAddress][votingOptionIndex];
+            }
+            //put the result into the voting result-array
+            result[votingOptionIndex] = amountVotes;
+        }
+        return result;
+    }
+
+    function formatedResults(uint256 _pollID) view public returns(FormatedResult[] memory) {
+        Poll storage poll = polls[_pollID];
+        uint256[] memory results = getResults(_pollID);
+        FormatedResult[] memory formatedResult = new FormatedResult[](poll.votingOptionsCount);
+
+        for(uint256 i = 0; i < poll.votingOptionsCount; i++) {
+            formatedResult[i] = FormatedResult(poll.votingOptions[i], results[i]);
+        }
+        return formatedResult;
+    }
+
+    function returnCoinsAfterPoll(uint256 _pollID) internal onlyOwner {
+        Poll storage poll = polls[_pollID];

```

Comment:
Should we require the poll's status to be `IN_RPOGRESS` / `CLOSED` (not sure)?

Reply:
Yup, macht Sinn